src/datasets/formatting/jax_formatter.py

```diff
@@ -15,25 +15,53 @@
 # Lint as: python3
 import sys
 from collections.abc import Mapping
-from typing import TYPE_CHECKING
+from typing import TYPE_CHECKING, Dict
 
 import numpy as np
 import pyarrow as pa
 
 from .. import config
+from ..utils.logging import get_logger
 from ..utils.py_utils import map_nested
 from .formatting import Formatter
 
 
 if TYPE_CHECKING:
     import jax
+    import jaxlib
+
+logger = get_logger()
 
 
 class JaxFormatter(Formatter[Mapping, "jax.Array", Mapping]):
-    def __init__(self, features=None, **jnp_array_kwargs):
+    def __init__(self, features=None, device=None, **jnp_array_kwargs):
         super().__init__(features=features)
+        import jax
+        from jaxlib.xla_client import Device
+
+        if isinstance(device, Device):
+            raise ValueError(
+                f"Expected {device} to be a `str` not {type(device)}, as `jaxlib.xla_extension.Device` "
+                "is not serializable neither with `pickle` nor with `dill`. Instead you can surround "
+                "the device with `str()` to get its string identifier that will be internally mapped "
+                "to the actual `jaxlib.xla_extension.Device`."
+            )
+        self.device = device if isinstance(device, str) else str(jax.devices()[0])
+        self.device_mapping = self._map_devices_to_str()
```

Comment:
wouldn't this end up make the formatter unpicklable ?

Reply:
So as far as I could test it was just unpickable when the `device` param is `jaxlib.xla_extension.Device`, otherwise it works fine. But let me check again maybe I skipped something?

Reply:
Indeed this works

```python
import jax
from datasets import Dataset
data = [[1, 2], [3, 4]]
ds = Dataset.from_dict({"data": data})
device = str(jax.devices()[0])
ds = ds.with_format("jax", device=device)
print(ds[0])
print(ds[0]["data"].device())
ds = ds.with_format("torch")
print(ds[0])
print(ds[0]["data"])
```

But if I remove the `ValueError` exception and instead replace that with the following:

```python
...
if isinstance(device, Device):
    device = str(device)
...
```

And then instead of the code above, I run the code below without the `str` casting:

```python
import jax
from datasets import Dataset
data = [[1, 2], [3, 4]]
ds = Dataset.from_dict({"data": data})
device = jax.devices()[0]
ds = ds.with_format("jax", device=device)
print(ds[0])
print(ds[0]["data"].device())
ds = ds.with_format("torch")
print(ds[0])
print(ds[0]["data"])
```

Then a warning will be raised on `with_format("jax", device=device)` and then `TypeError: cannot pickle 'jaxlib.xla_extension.Device' object` when calling `with_format("torch")`.

Reply:
Yea but I thought the dictionary values were `jaxlib.xla_extension.Device` objects ? Because of that the formatter itself is not picklable
```python
pickle.dumps(formatter)
```

I guess it's not possible to pass a string device to `jnp.array` ?

Reply:
So what I understand is that it should fail when `copy.deepcopy(self)` but it's not, so maybe the `copy.deepcopy` is done as part of the `super().__init__()` so that just the params are copied?

Reply:
And that's right, device management is done either with the context manager or with `device_put`, and it expects a `jaxlib.xla_extension.Device` not a string :/

Reply:
> So what I understand is that it should fail when copy.deepcopy(self) but it's not, so maybe the copy.deepcopy is done as part of the super().__init__() so that just the params are copied?

Actually we should have copy.deepcopy AND pickle.dumps working. That's because a formatter can also be used with other objects than a `Dataset`.

We could have the mapping as a global variable that is initialized when `_tensorize` is called for the first time ? This way we don't need to reload it every time we instantiate a formatter, and the formatter can become picklable

Reply:
Oh OK didn't know that, then I'll do what you're suggesting :)